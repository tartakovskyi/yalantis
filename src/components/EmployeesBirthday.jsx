import React  from 'react'


const EmployeesBirthday = () => {

  return (
    <div className="col-5">
    <h2>Employees Birthday</h2>
    </div>
    
  )
}


export default EmployeesBirthday