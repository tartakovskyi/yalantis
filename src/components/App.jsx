import React, { useEffect, useState }  from 'react'
import { apiGetUsers } from "../api"
import Employees from './Employees'
import EmployeesBirthday from './EmployeesBirthday'


export const AppContext = React.createContext()


const App = () => {

  const [users, setUsers] = useState({})

  useEffect(() => {
    apiGetUsers()
    .then(response => {
      let usersByAlphabet = {A:[],B:[],C:[],D:[],E:[],F:[],G:[],H:[],I:[],J:[],K:[],L:[],M:[],N:[],O:[],P:[],Q:[],R:[],S:[],T:[],U:[],V:[],W:[],X:[],Y:[],Z:[]}
      response.data.forEach(user => {
        const firstLetter = user.lastName.substr(0,1)
        usersByAlphabet[firstLetter].push(user)
      })
      setUsers(usersByAlphabet)
    })
    .catch(error => {
      console.log(error)
    })
  }, [])

  return (
    <div className="App">
      <header>
        <div className="container">
          <a href="/">Yalantis React School</a>
        </div>  
      </header>
      <main>
        <div className="container">    
          <div className="row">
            {Object.keys(users).length 
              ?
              <AppContext.Provider value={{users}}>
                <Employees />
                <EmployeesBirthday />    
              </AppContext.Provider>
              :
              null   
            }
          </div>
        </div>
      </main>
    </div>
  )
}

export default App
