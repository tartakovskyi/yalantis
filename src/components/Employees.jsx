import React, { useContext }  from 'react'
import { AppContext } from './App'
import Letter from './Letter'


const Employees = () => {

  const { users } = useContext(AppContext)
  const alphabet = ['A','B','C','D','E','F','G','H','I','J','K','L','M','N','O','P','Q','R','S','T','U','V','W','X','Y','Z']

  return (
    <div className="col-7">
      <h2>Employees</h2>
      <div className="row">
        {alphabet.map(letter => (<Letter letter={letter} users={users[letter]} key={letter} />))}
      </div>
    </div>
  )
}


export default Employees