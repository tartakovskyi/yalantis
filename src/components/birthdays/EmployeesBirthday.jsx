import React, { useContext, useEffect, useState } from 'react'
import { AppContext } from '../App'
import Month from './Month'


const EmployeesBirthday = () => {

  const { usersRaw, activeUsers } = useContext(AppContext)
  const [birthdaysByMonth, setBirthdaysByMonth] = useState({})
  const monthNames = ['January', 'February', 'March', 'April', 'May', 'June', 'July', 'August', 'September', 'October', 'November', 'December']

  useEffect(() => {
    let usersObj = {}
    usersRaw.forEach(user => {
      if (activeUsers.includes(user.id)) {
        const month = user.dob.replace(/([0-9]+)-([0-9]+)-(.*)/,'$2')
        if (usersObj.hasOwnProperty(month)) {
          usersObj[month].push(user)
        } else {
          usersObj[month] = [user]
        }
      }
    })
    setBirthdaysByMonth(usersObj)
  }, [usersRaw, activeUsers])

  const birthdaysArr = Object.keys(birthdaysByMonth).sort() 

  return (
    <div className='col-5'>
      <h2>Employees Birthday</h2>
      {birthdaysArr.length 
        ?
        <ul className="birthdays-list">
          {birthdaysArr.map(month => <Month name={monthNames[month - 1]} users={birthdaysByMonth[month]} key={month} />)}
        </ul>
        :
        <p>Employees List is empty</p>   
      }
    </div>
    
  )
}


export default EmployeesBirthday