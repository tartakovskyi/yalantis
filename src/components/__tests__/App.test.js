import React from "react";
import { render } from "@testing-library/react";
import App from "../App";
import Employees from "../employees/Employees";
import EmployeesBirthday from "../birthdays/EmployeesBirthday";

test("App should render correctly", () => {
  const { container, getByTestId } = render(<App />);

  expect(container.firstChild).toMatchSnapshot();
});
