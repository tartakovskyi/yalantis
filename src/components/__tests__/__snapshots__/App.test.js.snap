// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should render correctly 1`] = `
<div
  class="App"
>
  <header>
    <div
      class="container"
    >
      <a
        href="/"
      >
        Yalantis React.js School
      </a>
    </div>
  </header>
  <main>
    <div
      class="container"
    >
      <div
        class="row"
      />
    </div>
  </main>
</div>
`;
